src/lib.rs

```diff
@@ -69,75 +110,87 @@ pub unsafe extern "C" fn pthread_join(
-    ctru_sys::threadGetCurrent() as libc::pthread_t
-}
-
-unsafe fn get_main_thread_handle() -> Result<ctru_sys::Handle, ctru_sys::Result> {
-    // Unfortunately I don't know of any better way to get the main thread's
-    // handle, other than via svcGetThreadList and svcOpenThread.
-    // Experimentally, the main thread ID is always the first in the list.
-    let mut thread_ids = [0; 1];
-    let mut thread_ids_count = 0;
-    let result = ctru_sys::svcGetThreadList(
-        &mut thread_ids_count,
-        thread_ids.as_mut_ptr(),
-        thread_ids.len() as i32,
-        ctru_sys::CUR_PROCESS_HANDLE,
-    );
-    if ctru_sys::R_FAILED(result) {
-        return Err(result);
+    if let Some(thread) = THREAD_SELF {
+        return thread as libc::pthread_t;
     }
 
-    // Get the main thread's handle
-    let main_thread_id = thread_ids[0];
-    let mut main_thread_handle = 0;
-    let result = ctru_sys::svcOpenThread(
-        &mut main_thread_handle,
-        ctru_sys::CUR_PROCESS_HANDLE,
-        main_thread_id,
-    );
-    if ctru_sys::R_FAILED(result) {
-        return Err(result);
+    let thread = ctru_sys::threadGetCurrent();
+    let mut os_thread = 0;
+    let res = ctru_sys::svcGetThreadId(&mut os_thread, ctru_sys::CUR_THREAD_HANDLE);
+
+    if ctru_sys::R_FAILED(res) {
+        // TODO: some way to report error? This should probably normally never fail...
+        ptr::null_mut::<PThread>() as libc::pthread_t
+    } else {
+        let pthread = Box::new(PThread { thread, os_thread });
+        let raw = Box::into_raw(pthread);
+        THREAD_SELF = Some(raw);
+        raw as libc::pthread_t
     }
+}
+
+struct Handle(ctru_sys::Handle);
 
-    Ok(main_thread_handle)
+impl Handle {
+    fn is_valid(&self) -> bool {
+        self.0 != u32::MAX
+    }
 }
 
-unsafe fn get_thread_handle(native: libc::pthread_t) -> Result<ctru_sys::Handle, ctru_sys::Result> {
-    if native == MAIN_THREAD_ID {
-        get_main_thread_handle()
-    } else {
-        Ok(ctru_sys::threadGetHandle(native as ctru_sys::Thread))
+impl TryFrom<libc::pthread_t> for Handle {
+    type Error = ctru_sys::Result;
+
+    fn try_from(native: libc::pthread_t) -> Result<Self, Self::Error> {
+        // SAFETY:
+        // Clone the passed-in PThread to avoid a double-free when it gets joined or detached
+        unsafe {
+            let pthread = (&*(native as *const PThread)).clone();
+
+            let mut handle = 0;
+            let ret = ctru_sys::svcOpenThread(
+                &mut handle,
+                ctru_sys::CUR_PROCESS_HANDLE,
+                pthread.os_thread,
+            );
+
+            if ctru_sys::R_FAILED(ret) {
+                Err(ret)
+            } else {
+                Ok(Self(handle))
+            }
+        }
+    }
+}
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        unsafe {
+            let res = ctru_sys::svcCloseHandle(self.0);
+            assert!(ctru_sys::R_SUCCEEDED(res), "{:#X}", res);
```

Comment:
This is unsafe because it would unwind across an FFI boundary. I wouldn't worry about making sure the result is successful.

Reply:
Good point, probably just ignoring the error is easiest here. I think panicking here might normally abort anyway, so just as well.

Reply:
We have panics set to unwind, so it would trigger that whole mechanism